src/Transport/Rest/JsonStreamDecoder.php

```diff
@@ -0,0 +1,177 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore\Transport\Rest;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * JsonStreamDecoder is a HTTP-JSON response stream decoder for JSON-ecoded
+ * protobuf messages. The response stream must be a JSON array, where the first
+ * byte is the opening of the array (i.e. '['), and the last byte is the closing
+ * of the array (i.e. ']'). Each array item must be a JSON object and comma
+ * separated.
+ *
+ * The supported options include:
+ *     @type bool $ignoreUnknown
+ *           Toggles whether or not to throw an exception when an unknown field
+ *           is encountered in a response message. The default is true.
+ *     @type int $readChunkSizeBytes
+ *           The upper size limit in bytes that can be read at a time from the
+ *           response stream. The default is 1 KB.
+ *
+ * @experimental
+ */
+class JsonStreamDecoder
+{
+    const ESCAPE_CHAR = '\\';
+    private $stream;
+    private $decodeType;
+    private $ignoreUnknown = true;
+    private $readChunkSize = 1024;
+
+    public function __construct(StreamInterface $stream, $decodeType, $options = [])
+    {
+        $this->stream = $stream;
+        $this->decodeType = $decodeType;
+
+        if (!is_null($options)) {
+            $this->ignoreUnknown = array_key_exists('ignoreUnknown', $options) ?
+                                    $options['ignoreUnknown'] :
+                                    $this->ignoreUnknown;
+            $this->readChunkSize = array_key_exists('readChunkSizeBytes', $options) ?
+                                    $options['readChunkSizeBytes'] :
+                                    $this->readChunkSize;
+        }
+    }
+
+    /**
+     * Begins decoding the configured response stream. It is a generator which
+     * yields messages of the given decode type from the stream until the stream
+     * completes. Throws an Exception if the stream is closed before the closing
+     * byte is read or if it encounters an error while decoding a message.
+     *
+     * @throws Exception
+     * @return \Generator
+     */
+    public function decode()
+    {
+        $message = $this->decodeType;
+        $level = 0;
+        $str = false;
+        $prev = '';
+        $chunk = '';
+        $cursor = 0;
+        $start = 0;
+        $end = 0;
+        while ($chunk !== '' || !$this->stream->eof()) {
+            // Read up to $readChunkSize bytes from the stream.
+            $chunk .= $this->stream->read($this->readChunkSize);
```

Comment:
Is this `readChunkSize` fixed number? 
The data is sent in series of chunks, the length could be different. For [example](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#chunked_encoding),
```
HTTP/1.1 200 OK
Content-Type: text/plain
Transfer-Encoding: chunked

7\r\n
Mozilla\r\n
9\r\n
Developer\r\n
7\r\n
Network\r\n
0\r\n
\r\n
```
We should support random chunk size. But maybe I misreading here, correct me if I am wrong. :)

Reply:
It's a fixed number that is configurable by users and defaults to 1024 bytes. The code here will read messages that are split over the "chunk" boundary i.e. half of a message is in the first 1024 bytes, the next half of the message is in the next 1024 bytes. It could even be read 1 byte at a time. The chunk read size is really just a concern for performance.